### PR TITLE
Move source repartitioning into `ExecutionPlan::repartition`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/7935

## Rationale for this change

Inspired by PR https://github.com/apache/arrow-datafusion/pull/7745, the repartition input code as it is now is both a bit messy (needs a `#cfg`) but also only works for `ParquetExec` and `CsvExec`, where the code is general for any ExecutionPlan.

## What changes are included in this PR?

1. Extract the `repartition` code to a trait on ExecutionPlan, which also allows other (user defined) table sources to benefit from this optimization
2. Document what is happening in more detail
3. Move implementation code for Parquet and CSV

## Are these changes tested?

Covered by existing tests

### CODE REVIEW

File: datafusion/core/src/datasource/physical_plan/csv.rs [hunk 1]
```diff
@@ -117,34 +118,6 @@ impl CsvExec {
     pub fn escape(&self) -> Option<u8> {
         self.escape
     }
-
-    /// Redistribute files across partitions according to their size
-    /// See comments on `repartition_file_groups()` for more detail.
-    ///
-    /// Return `None` if can't get repartitioned(empty/compressed file).
-    pub fn get_repartitioned(
```
Comment (alamb) [inline datafusion/core/src/datasource/physical_plan/csv.rs:125]: This just was moved into the `impl ExecutionPlan` and the signature is changed

File: datafusion/core/src/physical_optimizer/enforce_distribution.rs [hunk 2]
```diff
@@ -26,9 +26,6 @@ use std::fmt::Formatter;
 use std::sync::Arc;
 
 use crate::config::ConfigOptions;
-use crate::datasource::physical_plan::CsvExec;
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/enforce_distribution.rs:29]: Now enforce distribution is not dependent on the specific operators 🎉

File: datafusion/physical-plan/src/lib.rs [hunk 3]
```diff
@@ -209,15 +210,34 @@ pub trait ExecutionPlan: Debug + DisplayAs + Send + Sync {
...
+    ///
+    /// If the `ExecutionPlan` does not support changing its partitioning,
+    /// returns `Ok(None)` (the default).
+    ///
+    /// The DataFusion optimizer attempts to use as many threads as possible by
+    /// repartitioning its inputs to match the target number of threads
+    /// available (`target_partitions`). Some data sources, such as the built in
+    /// CSV and Parquet readers, are able to read from their input files in
+    /// parallel, regardless of how the source data is split amongst files.
+    fn repartitioned(
```
Comment (alamb) [inline datafusion/physical-plan/src/lib.rs:224]: The PR hoists this common code to the `ExecutionPlan` trait, which is easier to understand as well as allowing user defined sources to be repartitioned, if they support that

**Comment by alamb** [comment]
@mustafasrepo  I wonder if you have time to review this PR

**Review by mustafasrepo — APPROVED** [review]
It is LGTM!. Thanks @alamb for this PR.

**Comment by alamb** [comment]
Thank you for the review @mustafasrepo

**Pull request merged**